projectq/backends/_azure/_utils.py

```diff
@@ -0,0 +1,307 @@
+)
+
+from .._exceptions import InvalidCommandError
+
+IONQ_PROVIDER_ID = 'ionq'
+QUANTINUUM_PROVIDER_ID = 'quantinuum'
+
+# https://docs.ionq.com/#section/Supported-Gates
+IONQ_GATE_MAP = {
+    HGate: 'h',
+    SGate: 's',
+    SqrtXGate: 'v',
+    SwapGate: 'swap',
+    TGate: 't',
+    Rx: 'rx',
+    Rxx: 'xx',
+    Ry: 'ry',
+    Ryy: 'yy',
+    Rz: 'rz',
+    Rzz: 'zz',
+    XGate: 'x',
+    YGate: 'y',
+    ZGate: 'z',
+}  # excluding controlled, conjugate-transpose and meta gates
+
+IONQ_SUPPORTED_GATES = tuple(IONQ_GATE_MAP.keys())
+
+QUANTINUUM_GATE_MAP = {
+    BarrierGate: 'barrier',
+    HGate: 'h',
+    Rx: 'rx',
+    Rxx: 'rxx',
+    Ry: 'ry',
+    Ryy: 'ryy',
+    Rz: 'rz',
+    Rzz: 'rzz',
+    SGate: 's',
+    TGate: 't',
+    XGate: 'x',
+    YGate: 'y',
+    ZGate: 'z',
+}  # excluding controlled, conjugate-transpose and meta gates
+
+QUANTINUUM_SUPPORTED_GATES = tuple(QUANTINUUM_GATE_MAP.keys())
+
+V = SqrtXGate()
+Vdag = get_inverse(V)
+
+
+def is_available_ionq(cmd):
+    """
+    Test if IonQ backend is available to process the provided command.
+
+    Args:
+        cmd (Command): A command to process.
+
+    Returns:
+        bool: If this backend can process the command.
+    """
+    gate = cmd.gate
+
+    if has_negative_control(cmd):
+        return False
+
+    if isinstance(gate, ControlledGate):
+        num_ctrl_qubits = gate._n  # pylint: disable=protected-access
+    else:
+        num_ctrl_qubits = get_control_count(cmd)
+
+    # Get base gate wrapped in ControlledGate class
+    if isinstance(gate, ControlledGate):
+        gate = gate._gate  # pylint: disable=protected-access
+
+    # NOTE: IonQ supports up to 7 control qubits
+    if 0 < num_ctrl_qubits <= 7:
+        return isinstance(gate, (XGate,))
+
+    # Gates without control bits
+    if num_ctrl_qubits == 0:
+        supported = isinstance(gate, IONQ_SUPPORTED_GATES)
+        supported_meta = isinstance(gate, (MeasureGate, AllocateQubitGate, DeallocateQubitGate))
+        supported_transpose = gate in (Sdag, Tdag, Vdag)
+
+        return supported or supported_meta or supported_transpose
+
+    return False
+
+
+def is_available_quantinuum(cmd):
+    """
+    Test if Quantinuum backend is available to process the provided command.
+
+    Args:
+        cmd (Command): A command to process.
+
+    Returns:
+        bool: If this backend can process the command.
+    """
+    gate = cmd.gate
+
+    if has_negative_control(cmd):
+        return False
+
+    if isinstance(gate, ControlledGate):
+        num_ctrl_qubits = gate._n  # pylint: disable=protected-access
+    else:
+        num_ctrl_qubits = get_control_count(cmd)
+
+    # Get base gate wrapped in ControlledGate class
+    if isinstance(gate, ControlledGate):
+        gate = gate._gate  # pylint: disable=protected-access
+
+    # TODO: NEEDED CONFIRMATION- Does Quantinuum support more than 2 control gates?
+    if 0 < num_ctrl_qubits <= 2:
+        return isinstance(gate, (XGate, ZGate))
```

Comment:
Shouldn't you also check the number of target qubits here? Or is there really no limits imposed by Quantinuum?

Reply:
It will take care by Azure platform on some level. For instance, if I run IonQ simulator with 30 qubits (maximum is 29), we will get below runtime error.

```
RuntimeError: Cannot retrieve results as job execution failed(status: Failed.error: {'additional_properties': {}, 'code': 'InsufficientResources', 'message': 'Too many qubits requested'})
```

I don't mind adding client-side validation as well. I happy to follow design principle used for ProjectQ backends.  Let me know your thoughts.

Reply:
No, that's ok. Let's not increase the amount of code we have to maintain too much. I will probably merge this over the weekend then.